base/linalg/bidiag.jl

```diff
@@ -41,26 +41,30 @@ julia> ev = [7, 8, 9]
  9
 
 julia> Bu = Bidiagonal(dv, ev, :U) # ev is on the first superdiagonal
-4×4 Bidiagonal{Int64}:
+4×4 Bidiagonal{Int64,Array{Int64,1}}:
  1  7  ⋅  ⋅
  ⋅  2  8  ⋅
  ⋅  ⋅  3  9
  ⋅  ⋅  ⋅  4
 
 julia> Bl = Bidiagonal(dv, ev, :L) # ev is on the first subdiagonal
-4×4 Bidiagonal{Int64}:
+4×4 Bidiagonal{Int64,Array{Int64,1}}:
  1  ⋅  ⋅  ⋅
  7  2  ⋅  ⋅
  ⋅  8  3  ⋅
  ⋅  ⋅  9  4
 ```
 """
-function Bidiagonal(dv::AbstractVector{T}, ev::AbstractVector{T}, uplo::Symbol) where T
-    Bidiagonal{T}(collect(dv), collect(ev), char_uplo(uplo))
+function Bidiagonal(dv::V, ev::V, uplo::Symbol) where {T,V<:AbstractVector{T}}
+    Bidiagonal{T}(dv, ev, char_uplo(uplo))
+end
+
+function Bidiagonal(dv::AbstractVector{T}, ev::AbstractVector{S}, uplo::Symbol) where {T,S}
+    R = promote_type(T,S)
+    Bidiagonal(convert(AbstractVector{R}, dv), convert(AbstractVector{R}, ev), uplo)
 end
 
-function Bidiagonal(dv::AbstractVector{Td}, ev::AbstractVector{Te}, uplo::Symbol) where {Td,Te}
-    T = promote_type(Td,Te)
+function Bidiagonal(dv::AbstractVector{T}, ev::AbstractVector{T}, uplo::Symbol) where T
     Bidiagonal(convert(Vector{T}, dv), convert(Vector{T}, ev), uplo)
```

Comment:
Thinking about this a bit more, perhaps we should not promote anything in the constructor here, and instead let the user promote to appropriate types. In that case this:

```julia
function Bidiagonal(dv::V, ev::V, uplo::Symbol) where {T,V<:AbstractVector{T}}
    Bidiagonal{T}(dv, ev, char_uplo(uplo))
end
```
 would be the only constructor. The good thing about this is that we don't try to "figure out" what the user wants; instead the user have to be explicit about this. The other good thing with this method is that the vectors in the `Bidiagonal` will *always* alias the vectors sent to the constructor. This would then be consistent with the new `Diagonal` type from #22718 . As this PR currently stands, the vectors will sometimes alias and sometimes not.